src/main/scala/gemmini/TilerFSM.scala

```diff
@@ -0,0 +1,718 @@
+//===========================================================================
+// TilerController's Internal FSM implementation
+//===========================================================================
+package gemmini
+
+import chisel3._
+import chisel3.util._
+import chisel3.experimental._
+import freechips.rocketchip.config._
+import freechips.rocketchip.tile._
+import GemminiISA._
+
+class TilerFSM[T <: Data : Arithmetic, U <: Data]
+  (config: GemminiArrayConfig[T,U])(implicit val p: Parameters) 
+  extends Module with HasCoreParameters {
+  import config._
+
+  //=========================================================================
+  // interface
+  //=========================================================================
+  val io = IO(new Bundle {
+    val cmd_in    = Flipped(Decoupled(new TilerCmd(LOG2_ROB_ENTRIES)))
+    val sched_out = Decoupled(new RoCCCommand)
+    val busy      = Output(Bool())
+  })
+  // hardcode 8 entries with up to 2 pushes per cycle
+  val schedq = Module(new MultiTailedQueue(new RoCCCommand, 8, 2))
+  io.sched_out <> schedq.io.deq
+
+  val sched = schedq.io.enq
+  val cmd   = io.cmd_in.bits
+  val busy  = io.busy
+
+  // initialize ports/pins
+  for (i <- 0 to 1) {
+    sched.bits(i) := DontCare
+    sched.bits(i).status := io.cmd_in.bits.status
+  }
+  sched.push := 0.U
+  io.cmd_in.ready := false.B
+  busy := true.B
+
+  //=========================================================================
+  // FSM states (see diagram for what each state does)
```

Comment:
Can you put a link to the diagram here?

Reply:
![tiler-fsm](https://user-images.githubusercontent.com/7218524/91516358-19979f80-e8a0-11ea-8869-df0dc88a8faa.png)